nameresolution/file/file.go

```diff
@@ -0,0 +1,190 @@
+		DaprPort:     metadata.Properties[nameresolution.DaprPort],
+		DaprHTTPPort: metadata.Properties[nameresolution.DaprHTTPPort],
+	}
+}
+
+type resolver struct {
+	logger logger.Logger
+	dir    string
+}
+
+// NewResolver creates file-based name resolver.
+func NewResolver(logger logger.Logger) nameresolution.Resolver {
+	if home, err := os.UserHomeDir(); err == nil {
+		return &resolver{
+			logger: logger,
+			dir:    filepath.Join(home, defaultDaprDir, defaultNamingDir),
+		}
+	}
+
+	return &resolver{
+		logger: logger,
+	}
+}
+
+// Init registers service for file-based name resolver.
+func (r *resolver) Init(metadata nameresolution.Metadata) error {
+	if err := r.prepareResolverDir(metadata); err != nil {
+		return err
+	}
+
+	info := createNamingInfo(metadata)
+
+	fileLock := r.newFileLock(info.AppID)
+	if err := fileLock.Lock(); err != nil {
+		r.logger.Errorf("fail to lock: %s", fileLock.Path())
+		return err
+	}
+
+	defer func(lck *flock.Flock) {
+		if err := lck.Unlock(); err != nil {
+			r.logger.Errorf("fail to unlock: %s, with error: %s", lck.Path(), err)
+		}
+	}(fileLock)
+
+	f := filepath.Join(r.dir, info.AppID)
+	namingInfos, err := loadNamingInfo(f)
+	if err != nil {
+		r.logger.Errorf("fail to load naming info from file: %s", f)
+		return err
+	}
+
+	index := -1
+	for i, ni := range namingInfos {
+		if ni.AppID == info.AppID &&
+			ni.HostAddress == info.HostAddress &&
+			ni.DaprPort == info.DaprPort &&
+			ni.DaprHTTPPort == info.DaprHTTPPort {
+			index = i
+			break
+		}
+	}
+
+	if index != -1 {
+		namingInfos = append(namingInfos[:index], namingInfos[index+1:]...)
+	}
+
+	namingInfos = append(namingInfos, *info)
+	content, _ := json.MarshalIndent(namingInfos, "", "\t")
+	if err := ioutil.WriteFile(f, content, os.ModePerm); err != nil {
+		r.logger.Errorf("fail to write naming info into file: %s", f)
+		return err
+	}
+
+	return nil
+}
+
+func loadNamingInfo(filename string) ([]namingInfo, error) {
+	_, err := os.Stat(filename)
+	if err != nil && os.IsNotExist(err) {
+		return make([]namingInfo, 0), nil
+	}
+
+	bytes, err := ioutil.ReadFile(filename)
+	if err != nil {
+		return nil, err
+	}
+
+	infos := make([]namingInfo, 0)
+	err = json.Unmarshal(bytes, &infos)
+	if err != nil {
+		return nil, err
+	}
+
+	return infos, nil
+}
+
+func (r *resolver) prepareResolverDir(metadata nameresolution.Metadata) error {
+	configs, err := config.Normalize(metadata.Configuration)
+	if err != nil {
+		return err
+	}
+
+	if conf, ok := configs.(map[string]interface{}); ok {
+		if v := conf[dirKey]; v != nil {
+			if s, ok := v.(string); ok {
+				r.dir = s
+			}
+		}
+	}
+
+	if r.dir == "" {
+		return errors.New("file based naming directory is not defined")
+	}
+
+	if err = os.MkdirAll(r.dir, os.ModePerm); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// ResolveID resolves name to address via file.
+func (r *resolver) ResolveID(req nameresolution.ResolveRequest) (string, error) {
+	lock := r.newFileLock(req.ID)
```

Comment:
Seems we need to read files when a request comes, how about reading one time then handle all route info, only re-read when the file changes?

Reply:
Can this be cached too? Maybe the file update happens in the background instead and the ResolveID() method will only do a lookup on a in-memory hash table.

Reply:
introduce filewatcher to load naming info only when necessary.